File: strategy-builder/components/agent/csharp/microsoft-agent-framework/KnowledgeBase.cs
namespace CairaAgent;

public sealed record KnowledgeEntry(string Id, string Title, string Summary, IReadOnlyList<string> Tags);

public static class KnowledgeBase
{
    private static readonly IReadOnlyList<KnowledgeEntry> Shanties =
    [
        new("shanty-stormcaller", "Stormcaller Refrain", "A rowing shanty with a thunder-and-tide rhythm that works well for dramatic openings.", ["storm", "rhythm", "opening", "chorus"]),
        new("shanty-bell", "Bell of the Watch", "A call-and-response chant sailors use when changing watches or challenging rivals.", ["watch", "call", "response", "challenge"]),
        new("shanty-harbor", "Harbor Lantern Song", "A lighter tune about returning to port, useful when praising a clever final verse.", ["harbor", "lantern", "praise", "return"])
    ];

    private static readonly IReadOnlyList<KnowledgeEntry> Treasures =
    [
        new("treasure-ruby", "Ruby Crown", "A jeweled crown rumored to rest in the Cavern of Echoes beneath a split reef.", ["ruby", "crown", "cavern", "reef"]),
        new("treasure-atlas", "Navigator Atlas", "A weathered atlas hidden in a shipwreck chapel and prized by captains seeking lost currents.", ["atlas", "shipwreck", "currents", "map"]),
        new("treasure-idol", "Whalebone Idol", "A carved idol guarded by tide pools and cliff paths on a fogbound island.", ["idol", "island", "fog", "cliff"])
    ];

    private static readonly IReadOnlyList<KnowledgeEntry> CrewRoles =
    [
        new("crew-quartermaster", "Quartermaster", "Keeps supplies, settles disputes, and favors recruits with practical judgment and discipline.", ["quartermaster", "supplies", "discipline", "judgment"]),
        new("crew-navigator", "Navigator", "Charts routes, reads stars, and suits recruits who stay calm and think ahead.", ["navigator", "stars", "planning", "calm"]),
        new("crew-lookout", "Lookout", "Spots hazards early and fits alert recruits with quick reactions and sharp eyes.", ["lookout", "alert", "hazards", "quick"])
    ];

    public static IReadOnlyList<KnowledgeEntry> LookupShanty(string query) => Lookup(Shanties, query);

    public static IReadOnlyList<KnowledgeEntry> LookupTreasure(string query) => Lookup(Treasures, query);

    public static IReadOnlyList<KnowledgeEntry> LookupCrew(string query) => Lookup(CrewRoles, query);

    private static IReadOnlyList<KnowledgeEntry> Lookup(IReadOnlyList<KnowledgeEntry> entries, string query)
    {
        var tokens = query.ToLowerInvariant()
            .Split([' ', ',', '.', ';', ':', '!', '?', '/', '-', '_'], StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);

        var ranked = entries
            .Select(entry => new
            {
                Entry = entry,
                Score = tokens.Length == 0
                    ? 1
                    : tokens.Count(token =>
                        entry.Title.Contains(token, StringComparison.OrdinalIgnoreCase) ||
                        entry.Summary.Contains(token, StringComparison.OrdinalIgnoreCase) ||
                        entry.Tags.Any(tag => tag.Contains(token, StringComparison.OrdinalIgnoreCase)))
            })
            .Where(item => item.Score > 0)
            .OrderByDescending(item => item.Score)
            .ThenBy(item => item.Entry.Title, StringComparer.Ordinal)
            .Take(3)
            .Select(item => item.Entry)
            .ToList();

        return ranked.Count > 0 ? ranked : entries.Take(2).ToList();
    }
}

File: strategy-builder/components/agent/typescript/foundry-agent-service/src/knowledge-base.ts
export interface KnowledgeEntry {
  readonly id: string;
  readonly title: string;
  readonly summary: string;
  readonly tags: readonly string[];
}

const SHANTY_ENTRIES: readonly KnowledgeEntry[] = [
  {
    id: 'shanty-stormcaller',
    title: 'Stormcaller Refrain',
    summary: 'A rowing shanty with a thunder-and-tide rhythm that works well for dramatic openings.',
    tags: ['storm', 'rhythm', 'opening', 'chorus']
  },
  {
    id: 'shanty-bell',
    title: 'Bell of the Watch',
    summary: 'A call-and-response chant sailors use when changing watches or challenging rivals.',
    tags: ['watch', 'call', 'response', 'challenge']
  },
  {
    id: 'shanty-harbor',
    title: 'Harbor Lantern Song',
    summary: 'A lighter tune about returning to port, useful when praising a clever final verse.',
    tags: ['harbor', 'lantern', 'praise', 'return']
  }
] as const;

const TREASURE_ENTRIES: readonly KnowledgeEntry[] = [
  {
    id: 'treasure-ruby',
    title: 'Ruby Crown',
    summary: 'A jeweled crown rumored to rest in the Cavern of Echoes beneath a split reef.',
    tags: ['ruby', 'crown', 'cavern', 'reef']
  },
  {
    id: 'treasure-atlas',
    title: 'Navigator Atlas',
    summary: 'A weathered atlas hidden in a shipwreck chapel and prized by captains seeking lost currents.',
    tags: ['atlas', 'shipwreck', 'currents', 'map']
  },
  {
    id: 'treasure-idol',
    title: 'Whalebone Idol',
    summary: 'A carved idol guarded by tide pools and cliff paths on a fogbound island.',
    tags: ['idol', 'island', 'fog', 'cliff']
  }
] as const;

const CREW_ENTRIES: readonly KnowledgeEntry[] = [
  {
    id: 'crew-quartermaster',
    title: 'Quartermaster',
    summary: 'Keeps supplies, settles disputes, and favors recruits with practical judgment and discipline.',
    tags: ['quartermaster', 'supplies', 'discipline', 'judgment']
  },
  {
    id: 'crew-navigator',
    title: 'Navigator',
    summary: 'Charts routes, reads stars, and suits recruits who stay calm and think ahead.',
    tags: ['navigator', 'stars', 'planning', 'calm']
  },
  {
    id: 'crew-lookout',
    title: 'Lookout',
    summary: 'Spots hazards early and fits alert recruits with quick reactions and sharp eyes.',
    tags: ['lookout', 'alert', 'hazards', 'quick']
  }
] as const;

function normalise(text: string): string[] {
  return text
    .toLowerCase()
    .split(/[^a-z0-9]+/)
    .filter(Boolean);
}

function scoreEntry(queryWords: readonly string[], entry: KnowledgeEntry): number {
  const haystack = [...entry.tags, entry.title, entry.summary].join(' ').toLowerCase();
  return queryWords.reduce((score, word) => (haystack.includes(word) ? score + 1 : score), 0);
}

function lookup(entries: readonly KnowledgeEntry[], query: string): readonly KnowledgeEntry[] {
  const words = normalise(query);
  const ranked = entries
    .map((entry) => ({ entry, score: words.length === 0 ? 1 : scoreEntry(words, entry) }))
    .filter((item) => item.score > 0)
    .sort((a, b) => b.score - a.score || a.entry.title.localeCompare(b.entry.title))
    .slice(0, 3)
    .map((item) => item.entry);

  return ranked.length > 0 ? ranked : entries.slice(0, 2);
}

export function lookupShantyKnowledge(query: string): readonly KnowledgeEntry[] {
  return lookup(SHANTY_ENTRIES, query);
}

export function lookupTreasureKnowledge(query: string): readonly KnowledgeEntry[] {
  return lookup(TREASURE_ENTRIES, query);
}

export function lookupCrewKnowledge(query: string): readonly KnowledgeEntry[] {
  return lookup(CREW_ENTRIES, query);
}
